packages/doenetml-worker-rust/doenetml-core/doenetml-derive/src/component_node.rs

```diff
@@ -0,0 +1,245 @@
+                            &self.public_state_variable_indices
+                        }
+
+                        fn get_state_variable_index_from_name(&self, name: &String) -> Option<usize> {
+                            self.state_variable_name_to_index.get(name).copied()
+                        }
+
+                        fn get_state_variable_index_from_name_case_insensitive(&self, name: &String) -> Option<usize> {
+                            self.state_variable_name_to_index
+                                .get_key_value_ignore_case(name)
+                                .map(|(k, v)| *v)
+                        }
+
+                        fn get_component_profile_state_variables(&self)  -> &Vec<ComponentProfileStateVariable> {
+                            &self.component_profile_state_variables
+                        }
+
+                        // fn get_essential_state_vars(&self) -> &HashMap<StateVarName, EssentialStateVar> {
+                        //     &self.essential_state_vars
+                        // }
+
+                    }
+                }
+            }
+            _ => panic!("only named fields supported"),
+        },
+        _ => panic!("only structs supported"),
+    };
+    output.into()
+}
+
+pub fn rendered_component_node_derive(input: TokenStream) -> TokenStream {
+    let ast: syn::DeriveInput = syn::parse(input).unwrap();
+    let name = &ast.ident;
+    let data = &ast.data;
+
+    let output = match data {
+        syn::Data::Struct(s) => match &s.fields {
+            syn::Fields::Named(FieldsNamed { .. }) => {
+                quote! {
+                    impl RenderedComponentNode for #name {
+                        // using default implementations for all traits so no code necessary here
+                    }
+                }
+            }
+            _ => panic!("only named fields supported"),
+        },
+        _ => panic!("only enums supported"),
+    };
+    output.into()
+}
+
+pub fn component_node_state_variables_derive(input: TokenStream) -> TokenStream {
+    let ast: syn::DeriveInput = syn::parse(input).unwrap();
+    let name = &ast.ident;
+    let data = &ast.data;
+
+    let output = match data {
+        syn::Data::Struct(s) => match &s.fields {
+            syn::Fields::Named(FieldsNamed { .. }) => {
+                quote! {
+                    impl ComponentNodeStateVariables for #name {
+                        // using default implementations for all traits so no code necessary here
+                    }
+                }
+            }
+            _ => panic!("only named fields supported"),
+        },
+        _ => panic!("only structs and enums supported"),
+    };
+    output.into()
+}
+
+pub fn add_standard_component_fields_impl(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let component_fields_to_add = [
+        quote! { pub idx: ComponentIdx},
+        quote! { pub parent: Option<ComponentIdx>},
```

Comment:
Strange spacing

Reply:
How would you want it spaced? The formatter is forcing this spacing.

Reply:
If the formatter is doing that, that's fine. It's just different fromt he lines below.

Reply:
Oh, I see what you are commenting on, the spacing inside the `quote!{}`. (I couldn't see the discrepancy as the snippet didn't show the lines below for comparison.) The formatter doesn't work inside the `quote!` macro, leading to such inconsistency if one isn't careful.  I'll fix that, assuming we want to keep this macro.